lib/src/value/calculation.dart

```diff
@@ -0,0 +1,387 @@
+// Copyright 2021 Google Inc. Use of this source code is governed by an
+// MIT-style license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+import 'package:meta/meta.dart';
+
+import '../exception.dart';
+import '../util/nullable.dart';
+import '../util/number.dart';
+import '../utils.dart';
+import '../value.dart';
+import '../visitor/interface/value.dart';
+import '../visitor/serialize.dart';
+
+/// A SassScript calculation.
+///
+/// Although calculations can in principle have any name or any number of
+/// arguments, this class only exposes the specific calculations that are
+/// supported by the Sass spec. This ensures that all calculations that the user
+/// works with are always fully simplified.
+///
+/// {@category Value}
+@sealed
+class SassCalculation extends Value {
+  /// The calculation's name, such as `"calc"`.
+  final String name;
+
+  /// The calculation's arguments.
+  ///
+  /// Each argument is either a [SassNumber], a [SassCalculation], an unquoted
+  /// [SassString], a [CalculationOperation], or a [CalculationInterpolation].
+  final List<Object> arguments;
+
+  /// @nodoc
+  @internal
+  bool get isSpecialNumber => true;
+
+  /// Creates a `calc()` calculation with the given [argument].
+  ///
+  /// The [argument] must be either a [SassNumber], a [SassCalculation], an
+  /// unquoted [SassString], a [CalculationOperation], or a
+  /// [CalculationInterpolation].
+  ///
+  /// This automatically simplifies the calculation, so it may return a
+  /// [SassNumber] rather than a [SassCalculation].
```

Comment:
Might be worth documenting that this throws if unsimplified args are not compatible (ditto for `min`, `max` and `clamp`)